File: apps/services/payments/src/flags.ts
// apps/services/payments/src/flags.ts
// Centralised prototype feature flags with typed accessors.

const TRUE_VALUES = new Set(["1", "true", "yes", "on"]);
const FALSE_VALUES = new Set(["0", "false", "no", "off"]);
const warnedInvalid = new Set<string>();

export type ProtoFlagName =
  | "PROTO_KILL_SWITCH"
  | "PROTO_ENABLE_IDEMPOTENCY"
  | "PROTO_ENABLE_RPT"
  | "PROTO_BLOCK_ON_ANOMALY"
  | "PROTO_ALLOW_OVERRIDES"
  | "PROTO_ENABLE_REAL_BANK";

export interface ProtoFlagsSnapshot {
  PROTO_KILL_SWITCH: boolean;
  PROTO_ENABLE_IDEMPOTENCY: boolean;
  PROTO_ENABLE_RPT: boolean;
  PROTO_BLOCK_ON_ANOMALY: boolean;
  PROTO_ALLOW_OVERRIDES: boolean;
  PROTO_ENABLE_REAL_BANK: boolean;
}

function parseBooleanFlag(name: ProtoFlagName, defaultValue: boolean): boolean {
  const raw = process.env[name];
  if (raw === undefined || raw.trim() === "") {
    return defaultValue;
  }

  const normalized = raw.trim().toLowerCase();
  if (TRUE_VALUES.has(normalized)) return true;
  if (FALSE_VALUES.has(normalized)) return false;

  if (!warnedInvalid.has(name)) {
    console.warn(`[flags] Ignoring invalid boolean for ${name}: ${raw}`);
    warnedInvalid.add(name);
  }
  return defaultValue;
}

export function getProtoFlags(): ProtoFlagsSnapshot {
  return {
    PROTO_KILL_SWITCH: parseBooleanFlag("PROTO_KILL_SWITCH", true),
    PROTO_ENABLE_IDEMPOTENCY: parseBooleanFlag("PROTO_ENABLE_IDEMPOTENCY", false),
    PROTO_ENABLE_RPT: parseBooleanFlag("PROTO_ENABLE_RPT", true),
    PROTO_BLOCK_ON_ANOMALY: parseBooleanFlag("PROTO_BLOCK_ON_ANOMALY", false),
    PROTO_ALLOW_OVERRIDES: parseBooleanFlag("PROTO_ALLOW_OVERRIDES", false),
    PROTO_ENABLE_REAL_BANK: parseBooleanFlag("PROTO_ENABLE_REAL_BANK", false),
  };
}

export function isProtoKillSwitchEnabled(): boolean {
  return getProtoFlags().PROTO_KILL_SWITCH;
}

export const PROTOTYPE_KILL_SWITCH_MESSAGE = "Prototype mode: egress disabled";

File: portal-api/flags.py
from __future__ import annotations

"""Shared prototype feature flag helpers for the portal API."""

from dataclasses import asdict, dataclass
import logging
import os
from typing import Dict

logger = logging.getLogger(__name__)

_TRUE_VALUES = {"1", "true", "yes", "on"}
_FALSE_VALUES = {"0", "false", "no", "off"}
_warned_invalid: set[str] = set()


def _parse_bool(name: str, default: bool) -> bool:
    raw = os.getenv(name)
    if raw is None or raw.strip() == "":
        return default

    normalized = raw.strip().lower()
    if normalized in _TRUE_VALUES:
        return True
    if normalized in _FALSE_VALUES:
        return False

    if name not in _warned_invalid:
        logger.warning("[flags] Ignoring invalid boolean for %s: %s", name, raw)
        _warned_invalid.add(name)
    return default


@dataclass(frozen=True)
class ProtoFlags:
    PROTO_KILL_SWITCH: bool
    PROTO_ENABLE_IDEMPOTENCY: bool
    PROTO_ENABLE_RPT: bool
    PROTO_BLOCK_ON_ANOMALY: bool
    PROTO_ALLOW_OVERRIDES: bool
    PROTO_ENABLE_REAL_BANK: bool


def get_proto_flags() -> ProtoFlags:
    return ProtoFlags(
        PROTO_KILL_SWITCH=_parse_bool("PROTO_KILL_SWITCH", True),
        PROTO_ENABLE_IDEMPOTENCY=_parse_bool("PROTO_ENABLE_IDEMPOTENCY", False),
        PROTO_ENABLE_RPT=_parse_bool("PROTO_ENABLE_RPT", True),
        PROTO_BLOCK_ON_ANOMALY=_parse_bool("PROTO_BLOCK_ON_ANOMALY", False),
        PROTO_ALLOW_OVERRIDES=_parse_bool("PROTO_ALLOW_OVERRIDES", False),
        PROTO_ENABLE_REAL_BANK=_parse_bool("PROTO_ENABLE_REAL_BANK", False),
    )


def get_proto_flags_dict() -> Dict[str, bool]:
    """Return the current flag snapshot as a serialisable dictionary."""

    return asdict(get_proto_flags())


def is_proto_kill_switch_enabled() -> bool:
    return get_proto_flags().PROTO_KILL_SWITCH


PROTOTYPE_KILL_SWITCH_MESSAGE = "Prototype mode: egress disabled"
